nixos/modules/services/databases/postgresql.nix

```diff
@@ -36,6 +43,104 @@ let
 
   groupAccessAvailable = versionAtLeast postgresql.version "11.0";
 
+  upgradeScript =
+    pkgs.writeShellApplication {
+      name = "upgrade.sh";
+      text = with cfg.upgradeFrom; ''
+        # There is no old version to upgrade from - noop
+        if ! [[ -d "${settings.old-datadir}" ]]; then
+          echo "No old datadir to upgrade from (expecting ${settings.old-datadir})"
+          exit 0
+        fi
+
+        if [[ -f "${settings.old-datadir}/.upgraded" ]]; then
+          echo "Old data dir found, but was already migrated. Please remove the old path '${settings.old-datadir}' to suppress this message." >&2
+          exit 0
+        fi
+
+        pushd "${settings.old-datadir}"
+        ${settings.new-bindir}/pg_upgrade ${lib.cli.toGNUCommandLineShell {} settings}
+        popd
+        touch "${settings.new-datadir}/.post_upgrade" "${settings.old-datadir}/.upgraded"
+      '';
+    };
+
+  upgradeOptions = upgrade: {
+    options = {
+      package = mkOption {
+        type = types.package;
+        default = pkgs."postgresql_${defaultPackageVersion}";
+        defaultText = literalExpression ''pkgs.postgresql_${defaultPackageVersion}'';
```

Comment:
I feel like it's fine because otherwise it would defeat the magical auto migration mechanism.

Reply:
> it would defeat the magical auto migration mechanism.

While this tooling is nice because it automates a lot (repetitive) stuff for you, but it should still be very obvious what's going on under the hood, so I guess I disagree I'm afraid.

Reply:
Your opinion is valuable and your are one of the maintainer, so I prefer we follow you here.